### ISSUE TITLE
AWS requests hangs after a few iterations.

### ISSUE THREAD

**Issue by skaggmannen** [aws/aws-xray-sdk-go#222]
I have written a small test app to try and reproduce issues we saw in our X-Ray enabled Lambdas when upgrading from 1.0.0-rc.14 to 1.0.0. 

Basically it seems that AWS requests hangs after 2-3 requests being made from the same Lambda instance causing the Lambdas to timeout. Exactly the same thing happens when running the following app:

```go
package main

import (
	"bytes"
	"encoding/json"
	"log"
	"net/http"
	"net/http/httptest"
	"os"

	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/dynamodb"
	"github.com/aws/aws-xray-sdk-go/xray"
	"github.com/aws/aws-xray-sdk-go/xraylog"
)

func main() {
	xray.SetLogger(xraylog.NewDefaultLogger(os.Stderr, xraylog.LogLevelDebug))

	sess := session.Must(session.NewSession())
	xray.AWSSession(sess)

	db := dynamodb.New(sess)

	var handler http.HandlerFunc = func(rsp http.ResponseWriter, req *http.Request) {
		log.Println("Listing tables...")
		output, err := db.ListTablesWithContext(req.Context(), &dynamodb.ListTablesInput{})
		if err != nil {
			panic(err)
		}

		rsp.WriteHeader(http.StatusOK)
		err = json.NewEncoder(rsp).Encode(output.TableNames)
		if err != nil {
			panic(err)
		}

		log.Println("Done!")
	}

	wrapped := xray.Handler(xray.NewFixedSegmentNamer("xray-test-app"), handler)

	for i := 0; i < 100; i++ {
		rsp := httptest.NewRecorder()
		req, err := http.NewRequest(http.MethodGet, "https://dummy/url", &bytes.Buffer{})
		if err != nil {
			panic(err)
		}

		wrapped.ServeHTTP(rsp, req)
		if rsp.Code != http.StatusOK {
			panic("expected ok")
		}
	}
}

```

Is this a bug introduced in 1.0.0 or am I doing something terribly wrong?

**Comment by bhautikpip** [comment]
Hi @skaggmannen ,

Thanks for opening an issue. I tried repro code that you have provided with v1.0.0 and I was not able to repro the issue you mentioned. I was able to send all the traces to service and process exited with 0 code. If you can share the screenshot of the error you're receiving we might be able to find the root cause. Also, I would recommend you to try out different aws sdk call and see if you can reproduce the same issue in the above code.

**Comment by skaggmannen** [comment]
I have a solid repro case on this using the v1.0.0 tag and the program above. 

I did some modifications to the test program to see stack dumps for all goroutines when the request hangs and found this:

`goroutine profile: total 8
2 @ 0x1037c60 0x10473c3 0x12d5adc 0x10647f1
#	0x12d5adb	net/http.(*persistConn).writeLoop+0x11b	/usr/local/Cellar/go/1.14/libexec/src/net/http/transport.go:2277

1 @ 0x1037c60 0x1007907 0x100760b 0x1529cc1 0x10647f1
#	0x1529cc0	github.com/aws/aws-xray-sdk-go/strategy/sampling.(*CentralizedStrategy).startRulePoller.func2+0x90	/Users/freall/go/pkg/mod/github.com/aws/aws-xray-sdk-go@v1.0.0/strategy/sampling/centralized.go:219

1 @ 0x1037c60 0x1007907 0x100760b 0x1529f2b 0x10647f1
#	0x1529f2a	github.com/aws/aws-xray-sdk-go/strategy/sampling.(*CentralizedStrategy).startTargetPoller.func1+0x8a	/Users/freall/go/pkg/mod/github.com/aws/aws-xray-sdk-go@v1.0.0/strategy/sampling/centralized.go:237

1 @ 0x1037c60 0x103132a 0x10308f5 0x10cd7a5 0x10ce6a1 0x10ce683 0x11ba20f 0x11cb9ae 0x12d32a5 0x116d1d3 0x116d33f 0x12d3f38 0x10647f1
#	0x10308f4	internal/poll.runtime_pollWait+0x54		/usr/local/Cellar/go/1.14/libexec/src/runtime/netpoll.go:203
#	0x10cd7a4	internal/poll.(*pollDesc).wait+0x44		/usr/local/Cellar/go/1.14/libexec/src/internal/poll/fd_poll_runtime.go:87
#	0x10ce6a0	internal/poll.(*pollDesc).waitRead+0x200	/usr/local/Cellar/go/1.14/libexec/src/internal/poll/fd_poll_runtime.go:92
#	0x10ce682	internal/poll.(*FD).Read+0x1e2			/usr/local/Cellar/go/1.14/libexec/src/internal/poll/fd_unix.go:169
#	0x11ba20e	net.(*netFD).Read+0x4e				/usr/local/Cellar/go/1.14/libexec/src/net/fd_unix.go:202
#	0x11cb9ad	net.(*conn).Read+0x8d				/usr/local/Cellar/go/1.14/libexec/src/net/net.go:184
#	0x12d32a4	net/http.(*persistConn).Read+0x74		/usr/local/Cellar/go/1.14/libexec/src/net/http/transport.go:1825
#	0x116d1d2	bufio.(*Reader).fill+0x102			/usr/local/Cellar/go/1.14/libexec/src/bufio/bufio.go:100
#	0x116d33e	bufio.(*Reader).Peek+0x4e			/usr/local/Cellar/go/1.14/libexec/src/bufio/bufio.go:138
#	0x12d3f37	net/http.(*persistConn).readLoop+0x1a7		/usr/local/Cellar/go/1.14/libexec/src/net/http/transport.go:1978

1 @ 0x1037c60 0x10473c3 0x12d472e 0x10647f1
#	0x12d472d	net/http.(*persistConn).readLoop+0x99d	/usr/local/Cellar/go/1.14/libexec/src/net/http/transport.go:2099

1 @ 0x1037c60 0x1047f00 0x1047eeb 0x1047c67 0x153bab1 0x153ba32 0x153defe 0x1406dc8 0x1409fd4 0x14098e7 0x1485dee 0x154efde 0x12bfc54 0x15327c9 0x153f6d0 0x12bfc54 0x154f205 0x10647f1
#	0x1047c66	sync.runtime_SemacquireMutex+0x46							/usr/local/Cellar/go/1.14/libexec/src/runtime/sema.go:71
#	0x153bab0	sync.(*RWMutex).RLock+0xa0								/usr/local/Cellar/go/1.14/libexec/src/sync/rwmutex.go:50
#	0x153ba31	github.com/aws/aws-xray-sdk-go/xray.(*Segment).getName+0x21				/Users/freall/go/pkg/mod/github.com/aws/aws-xray-sdk-go@v1.0.0/xray/segment.go:421
#	0x153defd	github.com/aws/aws-xray-sdk-go/xray.glob..func4+0x19d					/Users/freall/go/pkg/mod/github.com/aws/aws-xray-sdk-go@v1.0.0/xray/aws.go:112
#	0x1406dc7	github.com/aws/aws-sdk-go/aws/request.(*HandlerList).Run+0x97				/Users/freall/go/pkg/mod/github.com/aws/aws-sdk-go@v1.30.10/aws/request/handlers.go:267
#	0x1409fd3	github.com/aws/aws-sdk-go/aws/request.(*Request).sendRequest+0xb3			/Users/freall/go/pkg/mod/github.com/aws/aws-sdk-go@v1.30.10/aws/request/request.go:578
#	0x14098e6	github.com/aws/aws-sdk-go/aws/request.(*Request).Send+0x106				/Users/freall/go/pkg/mod/github.com/aws/aws-sdk-go@v1.30.10/aws/request/request.go:531
#	0x1485ded	github.com/aws/aws-sdk-go/service/dynamodb.(*DynamoDB).ListTablesWithContext+0xad	/Users/freall/go/pkg/mod/github.com/aws/aws-sdk-go@v1.30.10/service/dynamodb/api.go:3060
#	0x154efdd	main.main.func1+0xed									/tmp/xray-test/main.go:29
#	0x12bfc53	net/http.HandlerFunc.ServeHTTP+0x43							/usr/local/Cellar/go/1.14/libexec/src/net/http/server.go:2012
#	0x15327c8	github.com/aws/aws-xray-sdk-go/xray.httpTrace+0x518					/Users/freall/go/pkg/mod/github.com/aws/aws-xray-sdk-go@v1.0.0/xray/handler.go:147
#	0x153f6cf	github.com/aws/aws-xray-sdk-go/xray.Handler.func1+0x27f					/Users/freall/go/pkg/mod/github.com/aws/aws-xray-sdk-go@v1.0.0/xray/handler.go:115
#	0x12bfc53	net/http.HandlerFunc.ServeHTTP+0x43							/usr/local/Cellar/go/1.14/libexec/src/net/http/server.go:2012
#	0x154f204	main.main.func2+0x54									/tmp/xray-test/main.go:54

1 @ 0x12f3d85 0x12f3ba0 0x12f096a 0x154ee9f 0x1037882 0x10647f1
#	0x12f3d84	runtime/pprof.writeRuntimeProfile+0x94	/usr/local/Cellar/go/1.14/libexec/src/runtime/pprof/pprof.go:694
#	0x12f3b9f	runtime/pprof.writeGoroutine+0x9f	/usr/local/Cellar/go/1.14/libexec/src/runtime/pprof/pprof.go:656
#	0x12f0969	runtime/pprof.(*Profile).WriteTo+0x3d9	/usr/local/Cellar/go/1.14/libexec/src/runtime/pprof/pprof.go:329
#	0x154ee9e	main.main+0x56e				/tmp/xray-test/main.go:63
#	0x1037881	runtime.main+0x211			/usr/local/Cellar/go/1.14/libexec/src/runtime/proc.go:203

panic: timed out!

goroutine 1 [running]:
main.main()
	/tmp/xray-test/main.go:64 +0x58b
exit status 2
`

I believe the bug could be caused by the changes introduced in https://github.com/aws/aws-xray-sdk-go/commit/886d3aa07b41b88de2bb2cbade030779592b600c, more specifically in `xray/segment.go:291` where the function returns before `seg.Unlock()` is called.

**Comment by skaggmannen** [comment]
You can find the full source for the test program here: https://github.com/skaggmannen/xray-test

**Comment by bhautikpip** [comment]
Hi @skaggmannen ,

Sorry for the inconvenience. I was able to verify and reproduce this bug. You're right about the root cause of the bug too. Function should call seg.Unlock() before returning. I will create a PR for this fix as soon as possible. Thank you very much for this!

**Comment by duncancrawford** [comment]
@shengxil Any idea when this fix will be released?

**Comment by bhautikpip** [comment]
Hi @duncancrawford ,

We will be prioritizing release including this change soon. As of now I might not be able to share the exact release date.

**Comment by bhautikpip** [comment]
Hi ,

I recently released v1.0.1 (https://github.com/aws/aws-xray-sdk-go/releases/tag/v1.0.1) which includes fix for this issue (https://github.com/aws/aws-xray-sdk-go/pull/223). That should resolve most issues. Thanks for the patience! Resolving this issue now.
